src/connection/database/actions.js

```diff
@@ -61,41 +91,86 @@ export function signUp(id) {
       autoLogin: shouldAutoLogin(m)
     };
 
+    const isCaptchaValid = setCaptchaParams(m, params, fields);
+    if (!isCaptchaValid) {
+      return showMissingCaptcha(m, id);
+    }
+
     if (databaseConnectionRequiresUsername(m)) {
-      params.username = c.getFieldValue(m, 'username');
+      if (signUpHideUsernameField(m)) {
+        const usernameValidation = databaseConnection(m).getIn(['validation', 'username']);
+        const range = usernameValidation ? usernameValidation.toJS() : { max: 15 };
+        params.username = generateRandomUsername(range.max);
+      } else {
+        params.username = c.getFieldValue(m, 'username');
+      }
     }
     if (!additionalSignUpFields(m).isEmpty()) {
       params.user_metadata = {};
       params.user_metadata.extraFields = {};
       additionalSignUpFields(m).forEach(x => {
-        if (x.get('type') !== 'textinfo') {
-          if (x.get('isExtra')) {
-            params.user_metadata.extraFields[x.get('name')] = c
-              .getFieldValue(m, x.get('name'))
-              .trim();
-          } else {
-            params.user_metadata[x.get('name')] = c.getFieldValue(m, x.get('name')).trim();
-          }
+        const storage = x.get('storage');
+        const fieldName = x.get('name');
+        const fieldValue = c.getFieldValue(m, x.get('name')).trim();
+        const isExtra = x.get('isExtra')
+        switch (storage) {
+          case 'root':
+            params[fieldName] = fieldValue;
```

Comment:
isn't there a risk that text info fields are stored here by mistake?

Reply:
Added the check for textinfo.